### ISSUE TITLE
Warnings re rb_tainted_str_new_cstr in MRI 2.7.0

### ISSUE THREAD

**Issue by gdtrfb55** [robinst/taglib-ruby#86]
A code segment using taglib-ruby 0.7.1 in a private project of mine works fine with MRI 2.6.5. However, when I run the same code with MRI 2.7.0, I get this warning (repeated multiple times):

`/home/jack/bin/oggidx:55: warning: rb_tainted_str_new_cstr is deprecated and will be removed in Ruby 3.2.`

The ruby code being called at line 55 is this:

`TagLib::Ogg::Vorbis::File.open(ogg) { |f| f.tag.field_list_map }`

The code runs without error under both versions, but the avalanche of warnings produced with 2.7.0 is problematic. I can work around the warnings (by ignoring them), but it seems there is an issue here (at least when MRI 3.2 rolls around).

Note: the code was run under both versions with $VERBOSE set to 'true'.

**Comment by jacobvosmaer** [comment]
The gem should not cause those warnings. I suspect they stem from https://github.com/robinst/taglib-ruby/blob/master/ext/taglib_base/includes.i .

If we find out what Ruby 2.7 wants us to do instead we can consider options for addressing this. Either use a different API, or use ifdefs to detect Ruby 2.7 and up at compilation time.

**Comment by gdtrfb55** [comment]
Roger that. Definitely a 2.7 issue, because taglib-ruby 0.7.1 (and earlier) ran without incident on MRI < 2.7.0. I worked around the current issue with this kludge:

````
unless Kernel.respond_to? :silence_warnings
  module Kernel
    def with_warnings(flag)
      old_verbose = $VERBOSE
      $VERBOSE = flag
      yield
    ensure
      $VERBOSE = old_verbose
    end

    def silence_warnings
      with_warnings(nil) { yield }
    end
  end
end

I then wrapped the call to taglib-ruby with:

silence_warnings do
  TagLib::Ogg::Vorbis::File.open(ogg) { |f| f.tag.field_list_map }
end

This is just a band-aid, though.

**Comment by gdtrfb55** [comment]
I poked around the Ruby git and the ruby-core mailing list, but to no avail.

**Comment by jacobvosmaer** [comment]
https://github.com/ruby/ruby/blob/master/NEWS#other-miscellaneous-changes

Object#{taint,untaint,trust,untrust} and related functions in the C-API no longer have an effect (all objects are always considered untainted), and are now warned in verbose mode. This warning will be disabled even in non-verbose mode in Ruby 3.0, and the methods and C functions will be removed in Ruby 3.2. [Feature #16131]

https://bugs.ruby-lang.org/issues/16131

**Comment by jacobvosmaer** [comment]
If I read that ruby "bug" issue correctly, the idea is that the "taint" mechanism is considered more or less pointless. An idea inherited from Perl from the CGI era.

@robinst if that is so we should probably just switch to regular (untainted) strings in taglib.

**Comment by gdtrfb55** [comment]
Interesting. I ran across a big patch in November that noop-ed (with warning) a bunch of rb_tainted_str... functions, including rb_tainted_str_new_cstr. That must have been the patch implementing the change described in 16131. Looks like that's the issue.

**Comment by gdtrfb55** [comment]
According to lines 911-922 of [string.c](https://github.com/ruby/ruby/blob/master/string.c), _rb_tainted_str_new_ and _rb_tainted_str_new_cstr_ now just issue a deprecation warning and call _rb_str_new_ and _rb_str_new_cstr_, respectively. So, substituting regular string calls for their tainted counterparts would achieve the same functionality without the warnings.

**Comment by gdtrfb55** [comment]
Line 907 of [intern.h](https://github.com/ruby/ruby/blob/master/include/ruby/intern.h) aliases _rb_tainted_str_new2_ to _rb_tainted_str_new_cstr_, so replacing that with its regular version, _rb_str_new2_ (which is itself an alias of _rb_str_new_cstr_), would seem to be OK, too.

**Comment by jacobvosmaer** [comment]
I tried perparing a PR with this change but I got stuck because I don't have a working build environment. The existing wrappers were generated with Swig 3.0.7 but the environment I managed set up has 3.0.12. Now I get a crash in one of the tests:

```
$ bundle exec ruby -I. test/file_test.rb 
Loaded suite test/file_test
Started
..
Finished in 0.000606954 seconds.
-------------------------------------------------------------------------------------------
2 tests, 2 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
-------------------------------------------------------------------------------------------
3295.14 tests/s, 3295.14 assertions/s
Aborted (core dumped)
```

This seems to be orthogonal to the tainted string change. I get this just by regenerating the wrappers with Swig 3.0.12.

```
rm $(git ls-files|grep wrap.cxx)
rake clean swig compile
bundle exec ruby -I. test/file_test.rb 
```

So either we fix that, or whoever submits the patch needs to set up Swig 3.0.7 in their development environment. That "whoever" might be me but not today.

**Comment by Sigill** [comment]
I think I'm the one who last generated the wrappers (https://github.com/robinst/taglib-ruby/pull/83).
The current release of Swig at that time was 3.0.12, but it had a memory management issue when generating in Ruby wrappers since 3.0.8. I therefore used 3.0.7.
I submitted a patch to Swig (https://github.com/swig/swig/pull/1177/commits/a4884e45e1aa10623d03d38eaae167ca6b6eb665), but it's only available from Swig 4.0.0.
I think that building Swig was not that difficult as long as you only enable Ruby in the supported languages (during configure?). Once built, just add it to your PATH.

**Comment by jacobvosmaer** [comment]
@Sigill thanks! I managed to create #89.
